### PR TITLE
Remaining code changes for Dart 2.0 SDK roll.

### PR DESCRIPTION
**Pull request by leafpetersen**
This is the remainder of the code changes required for rolling flutter to the Dart 2.0 SDK.  The actual  SDK + package roll will follow in a separate PR.  Changes in this PR:

- One Map.fromIterable to "for in" loop change, on new code added since the last PR
- Add "AnyDistanceFunction" typedef to describe union of different DistanceFunctions
- Make the Navigator.push static method generic
- Make the NavigatorState.push instance method use Object instead of dynamic to avoid bad interaction between "--no-implicit-dynamic" and "@optionalTypeArgs"

### CODE REVIEW

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -913,7 +914,7 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   ///
   /// Returns a [Future] that completes to the `result` value passed to [pop]
   /// when the pushed route is popped off the navigator.
-  Future<dynamic> push(Route<dynamic> route) {
+  Future<Object> push(Route<Object> route) {
```
Comment (leafpetersen) [inline packages/flutter/lib/src/widgets/navigator.dart]: I had originally made this generic as well, but "--no-implicit-dynamic" doesn't know about @optionalTypeArgs, and so some explicit `push<dynamic>` calls were required.  Using `Object` here avoids introducing `dynamic` at all.

File: packages/flutter_test/lib/src/matchers.dart [hunk 2]
```diff
@@ -592,7 +592,19 @@ class _HasGoodToStringDeep extends Matcher {
 /// sets of value for which a metric space is defined.
 typedef num DistanceFunction<T>(T a, T b);
 
-const Map<Type, DistanceFunction<dynamic>> _kStandardDistanceFunctions = const <Type, DistanceFunction<dynamic>>{
+/// The type of a union of instances of DistanceFunction<T> for various types T.
```
Comment (leafpetersen) [inline packages/flutter_test/lib/src/matchers.dart]: @Hixie This is the change we discussed when I was last in MTV.  Let me know if you have better suggestions for the comment wording, or would prefer to mediate through this through a generic `DistanceFunction<T> getDistanceFunction<T>(Type t)` API.

**Comment by leafpetersen** [comment]
cc @mit-mit

**Comment by leafpetersen** [comment]
New Year's ping!  :)

File: packages/flutter_test/lib/src/matchers.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_test/lib/src/matchers.dart]: `[DistanceFunction<T>]`

File: packages/flutter_test/lib/src/matchers.dart [hunk 3]
```diff
@@ -592,7 +592,19 @@ class _HasGoodToStringDeep extends Matcher {
...
+///
+/// This type is used to describe a collection of DistanceFunction<T> functions
```
Comment (Hixie) [inline packages/flutter_test/lib/src/matchers.dart]: ditto

**Comment by Hixie** [comment]
LGTM

File: packages/flutter_test/lib/src/matchers.dart (hunk 2 above)
Comment (leafpetersen) [reply packages/flutter_test/lib/src/matchers.dart]: Done.

File: packages/flutter_test/lib/src/matchers.dart (hunk 3 above)
Comment (leafpetersen) [reply packages/flutter_test/lib/src/matchers.dart]: Done.

**Pull request merged**